src/v/iceberg/tests/schema_protobuf_test.cc

```diff
@@ -0,0 +1,187 @@
+#include <google/protobuf/descriptor.h>
+#include <google/protobuf/descriptor.pb.h>
+#include <google/protobuf/io/tokenizer.h>
+#include <gtest/gtest.h>
+
+#include <memory>
+using namespace iceberg;
+
+namespace {
+
+void parse_proto_definition_string(
+  const ss::sstring& proto_str, google::protobuf::DescriptorPool& pool) {
+    google::protobuf::compiler::Parser parser;
+
+    google::protobuf::io::ArrayInputStream input_stream(
+      proto_str.c_str(), proto_str.size());
+
+    google::protobuf::io::Tokenizer tokenizer(&input_stream, nullptr);
+    google::protobuf::FileDescriptorProto fdp;
+    fdp.set_name("test_schema_file");
+
+    auto parse_result = parser.Parse(&tokenizer, &fdp);
+    ASSERT_TRUE(parse_result);
+    pool.BuildFile(fdp);
+}
+
+}; // namespace
+
+MATCHER_P3(IsField, id, name, type, "") {
+    *result_listener << fmt::format(
+      "field(id: {}, name: {}, type: {}) expected to be equal to "
+      "field(id: {}, "
+      "name: {}, type: {})\n",
+      arg->id,
+      arg->name,
+      arg->type,
+      id,
+      name,
+      type);
+
+    return arg->id == id && arg->name == name && arg->type == type;
+}
+
+TEST(SchemaProtobuf, TestSimpleSchemaConversion) {
+    google::protobuf::DescriptorPool d_pool;
+    parse_proto_definition_string(basic_proto_str, d_pool);
+    auto person_descriptor = d_pool.FindMessageTypeByName("Person");
+
+    auto result = iceberg::type_to_iceberg(*person_descriptor);
+    ASSERT_FALSE(result.has_error());
+    auto& top_level = result.value();
+    // string name = 1;
+    EXPECT_THAT(top_level.fields[0], IsField(0, "name", string_type{}));
+    // string id = 2;
+    EXPECT_THAT(top_level.fields[1], IsField(1, "id", int_type{}));
+    // string email = 3;
+    EXPECT_THAT(top_level.fields[2], IsField(2, "email", string_type{}));
+
+    // Department dept = 4; (nested struct)
+    EXPECT_EQ(top_level.fields[3]->id, 3);
+    EXPECT_EQ(top_level.fields[3]->name, "dept");
+    auto& nested = std::get<struct_type>(top_level.fields[3]->type);
+    // int32 id = 1;
+    // string name = 2;
+    EXPECT_THAT(nested.fields[0], IsField(0, "id", int_type{}));
+    EXPECT_THAT(nested.fields[1], IsField(1, "name", string_type{}));
+    // string test_coverage = 5;
+    EXPECT_THAT(
+      top_level.fields[4], IsField(4, "test_coverage", string_type{}));
+}
+
+/**
+ *enum State {
+ *  Active = 0;
+ *  InActive = 1;
+ *}
+ *
+ *message BrokerShard {
+ *  int64 id = 1;
+ *  int32 shard = 2;
+ *}
+ *
+ *message Partition {
+ *    int32 id = 1;
+ *    repeated BrokerShard replicas = 2;
+ *    State state = 3;
+ *}
+ *
+ *message Topic {
+ *    string name = 1;
+ *    repeated Partition partitions = 2;
+ *}
+ *
+ *message Metadata {
+ *    map<string,Topic> topics = 1;
+ *    repeated int64 nodes = 2;
+ *}
+ */
+// this schema contains a list map and nested fields
+// static_assert(
+// SchemaConverter<protobuf_schema_converter, google::protobuf::Descriptor>);
+TEST(SchemaProtobuf, TestComplexSchema) {
```

Comment:
In //src/v/serde/protobuf/tests I have a test proto from the official conformance tests that contains almost all protobuf types (not all 72 variants of maps but a lot of them). Can we use that for exhaustiveness?

Reply:
That is awesome 🔥, thank you for pointing me to that.